src/cascadia/TerminalApp/TerminalPage.cpp

```diff
@@ -2363,13 +2341,43 @@ namespace winrt::TerminalApp::implementation
         return _isAlwaysOnTop;
     }
 
-    void TerminalPage::_OnNewConnection(winrt::Microsoft::Terminal::TerminalConnection::ITerminalConnection connection)
+    HRESULT TerminalPage::_OnNewConnection(winrt::Microsoft::Terminal::TerminalConnection::ITerminalConnection connection)
     {
-        // TODO: GH 9458 will give us more context so we can try to choose a better profile.
-        _OpenNewTab(nullptr, connection);
+        // We need to be on the UI thread in order for _OpenNewTab to run successfully.
+        // HasThreadAccess will return true if we're currently on a UI thread and false otherwise.
+        // When we're on a COM thread, we'll need to dispatch the calls to the UI thread
+        // and wait on it hence the locking mechanism.
+        if (Dispatcher().HasThreadAccess())
+        {
+            // TODO: GH 9458 will give us more context so we can try to choose a better profile.
+            auto hr = _OpenNewTab(nullptr, connection);
+
+            // Request a summon of this window to the foreground
+            _SummonWindowRequestedHandlers(*this, nullptr);
 
-        // Request a summon of this window to the foreground
-        _SummonWindowRequestedHandlers(*this, nullptr);
+            return hr;
+        }
+        else
+        {
+            std::optional<HRESULT> finalVal;
+            std::mutex mtx;
+            std::condition_variable cv;
+
+            std::unique_lock<std::mutex> lock{ mtx };
+
+            Dispatcher().RunAsync(CoreDispatcherPriority::Normal, [&]() {
+                auto hr = _OpenNewTab(nullptr, connection);
+
+                _SummonWindowRequestedHandlers(*this, nullptr);
+
+                std::unique_lock<std::mutex> lock{ mtx };
```

Comment:
Something is up here and I'm not certain what it is.

It worries me that you covered one definition of lock (since the lambda captured everything) with another more local definition of the same thing.

The inside of this lambda isn't even using the lock. Is it? Am I missing an implicit piece of the construction here?

Reply:
I should probably change the name of the lock inside of the lambda, that's a good point.
As for whether the inside of the lambda is using the lock, we should technically be acquiring the lock before modifying `finalVal` since that's the shared resource right?

Reply:
I need an adult to answer that. I just copied this pattern from somewhere else. @lhecker halp.

Reply:
(This is where I stole the idea from: https://github.com/microsoft/terminal/blob/d57fb84557e993017a2fecc346ae62cef1803962/src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp#L57-L78)

Reply:
You can use [`til::latch`](https://github.com/microsoft/terminal/blob/e7108332f72a3c14c72a65484d091c4e79ab28b2/src/inc/til/latch.h) for this now. Less overhead & easier to use. 🙂
* You need to import it yourself as `<til/latch.h>` (just like a stdlib header - no need to put it into `pch.h`)
* Instantiate it as `til::latch latch(1);`
* Replace the mutex/condition_variable code inside the lambda with `lath.count_down()`
* Replace the mutex/condition_variable code outside the lambda with `latch.wait()`
* Just make sure to write the data before counting down and only read from it after wait() has returned.